### ISSUE TITLE
6.0.12: SagePay (Form), people from the US cannot pay

### ISSUE THREAD

**Issue by cb2004** [cubecart/v6#1220]
Using SagePay (Form), people from the US cannot pay. It seems this block of code tries to do something but it is not working:

```
if($this->_basket['delivery_address']['country_iso']=="US") {
	if(strlen($this->_basket['billing_address']['state_abbrev']) > 2) {
		$this->_basket['billing_address']['state_abbrev'] = getStateFormat($this->_basket['billing_address']['state_abbrev'], 'name', 'abbrev');
	}
	$cryptVars .= "&BillingState=" .$this->_ci($this->_basket['billing_address']['state_abbrev'], CLEAN_INPUT_FILTER_TEXT);
}
if($this->_basket['delivery_address']['country_iso']=="US") {
	if(strlen($this->_basket['delivery_address']['state_abbrev']) > 2) {
		$this->_basket['delivery_address']['state_abbrev'] = getStateFormat($this->_basket['delivery_address']['state_abbrev'], 'name', 'abbrev');
	}
	$cryptVars .= "&DeliveryState=" .$this->_ci($this->_basket['delivery_address']['state_abbrev'],CLEAN_INPUT_FILTER_TEXT);
}
```

**Comment by cb2004** [comment]
This changed code seems to do the trick. Note that the first if statement had 2 checks for delivery address, there are no abbreviation lengths greater than 2 in my store so I commented out that and it seems to work. Probably not the ideal fix but got it working.

```
if($this->_basket['billing_address']['country_iso']=="US") {
	//if(strlen($this->_basket['billing_address']['state_abbrev']) > 2) {
	//	$this->_basket['billing_address']['state_abbrev'] = getStateFormat($this->_basket['billing_address']['state_abbrev'], 'name', 'abbrev');
	//}
	$cryptVars .= "&BillingState=" .$this->_ci($this->_basket['billing_address']['state_abbrev'], CLEAN_INPUT_FILTER_TEXT);
}
if($this->_basket['delivery_address']['country_iso']=="US") {
	//if(strlen($this->_basket['delivery_address']['state_abbrev']) > 2) {
	//	$this->_basket['delivery_address']['state_abbrev'] = getStateFormat($this->_basket['delivery_address']['state_abbrev'], 'name', 'abbrev');
	//}
	$cryptVars .= "&DeliveryState=" .$this->_ci($this->_basket['delivery_address']['state_abbrev'],CLEAN_INPUT_FILTER_TEXT);
}
```

**Comment by cb2004** [comment]
@abrookbanks just a heads up that this is obviously a critical issue and I dont see an updated version in the marketplace yet, my fix is working but I am wondering if this is the best fix. Luckily my orders from the US are now moving to processing but there was a lot of lost money for this one particular store, over 20 orders from the US sitting at pending since the launch of the version 6 site a couple of months back.

**Comment by abrookbanks** [comment]
I haven't looked at this yet. I simply have to prioritise my time on tech support and I'll get onto this when I can. 

> Using SagePay (Form), people from the US cannot pay.

Do you get an error message from SagePay? Might be a silly question but in what why can't they pay?

**Comment by cb2004** [comment]
@abrookbanks I fully understand, that is the difficulty with the mix of open source and paid support. I just wanted to highlight it again.

There was an error message and number, I cant remember which one it was now though, something along the lines of malformed. However if I un-comment the lines that I commented out it seems ok now. Maybe this was an error on the sage side, although the first if statement should check for billing address and not delivery shouldnt it?

**Comment by abrookbanks** [comment]
> There was an error message and number, I cant remember which one it was now though, something along the lines of malformed.

If you can find out it will help greatly.

**Comment by cb2004** [comment]
@abrookbanks yeah run a load of tests this morning, its just not throwing it. I have told the store owner to keep an eye out for any US orders that stick at pending. I am now using the default module code again just with the if statement changed.

**Comment by abrookbanks** [comment]
I wonder if the state_abbrev in your store are bad?

California for example should be CA. Anything over 2 chars may cause this issue. Can you check that? Look at the Countries/Zones section of ACP.

**Comment by cb2004** [comment]
@abrookbanks no they are fine, they are one of the things I went in and checked when originally trying to debug. When you enter that page in the admin does it cache anything? I would have never gone into this section after the upgrade from V4.

**Comment by abrookbanks** [comment]
I can't reproduce this. 

Movie: http://static.cubecart.com/github-1220.mp4

**Comment by cb2004** [comment]
@abrookbanks yes, this seems to be a problem that Sage were having. I see you have release a 1.0.4 but the if statement has still not been fixed. Line 255 should be:

`if($this->_basket['billing_address']['country_iso']=="US") {`

**Comment by abrookbanks** [comment]
Using the latest version I am not able to reproduce this problem. The movie is 1.0.4.

**Comment by cb2004** [comment]
@abrookbanks I agree, I cannot reproduce with my tests either so it was a problem with Sage. What I am saying is there is this block of code starting at line 255 in 1.0.4:

```
if($this->_basket['delivery_address']['country_iso']=="US") {
	if(strlen($this->_basket['billing_address']['state_abbrev']) > 2) {
		$this->_basket['billing_address']['state_abbrev'] = getStateFormat($this->_basket['billing_address']['state_abbrev'], 'name', 'abbrev');
	}
	$cryptVars .= "&BillingState=" .$this->_ci($this->_basket['billing_address']['state_abbrev'], CLEAN_INPUT_FILTER_TEXT);
}
if($this->_basket['delivery_address']['country_iso']=="US") {
	if(strlen($this->_basket['delivery_address']['state_abbrev']) > 2) {
		$this->_basket['delivery_address']['state_abbrev'] = getStateFormat($this->_basket['delivery_address']['state_abbrev'], 'name', 'abbrev');
	}
	$cryptVars .= "&DeliveryState=" .$this->_ci($this->_basket['delivery_address']['state_abbrev'],CLEAN_INPUT_FILTER_TEXT);
}
```

There is a double check for delivery_address. The first check should be for billing_address

**Comment by abrookbanks** [comment]
Thanks! I released 1.0.5.
